### ISSUE TITLE
Add support of wasm-bindgen

### ISSUE THREAD

**Issue by DmitryAstafyev** [DmitryAstafyev/tslink#3]
Add support of `wasm-bindgen`:
- [ ] prevent conflict between js wrapper created by `tslink` and `wasm-bindgen`
- [ ] add (optionally) loader of wasm module
- [ ] add tests
- [ ] add examples
- [ ] update README

**Comment by icsmw** [comment]
@siefkenj I would suggest continue discussion of `wasm-bindgen` here

**Comment by icsmw** [comment]
Hello @nappa85, maybe this crate could be interested for you too. If you have some time, it would be nice to hear: what is missed on `wasm-bindgen` from typescript perspective. Welcome to discussion. 
@nappa85 for you - `tslink` is one more crate, which allows represent types from rust to TS

**Comment by nappa85** [comment]
Am I wrong or this isn't serde-based? Tsify is good because it's
serde-based, you just need to add a derive macro to an existing codebase,
where serde is already setup for your structs, and it's already fully
integrated with wasm-bindgen

Il sab 13 gen 2024, 23:21 icsmw ***@***.***> ha scritto:

**Comment by icsmw** [comment]
> Am I wrong or this isn't serde-based? Tsify is good because it's serde-based, you just need to add a derive macro to an existing codebase, where serde is already setup for your structs, and it's already fully integrated with wasm-bindgen Il sab 13 gen 2024, 

@nappa85 yes, if we are talking about not primitive types, it can be converted to JSON based on serde indeed. But I'm wondering in general, what is missed in `wasm-bindgen` out of the box? Because out of the box `wasm-bindgen` already creates `.d.ts` files. Is the only problem with not primitive types, which are represented in TypeScript as `any`?

**Comment by nappa85** [comment]
Main problem with wasm-bindgen is it can't expose generics and enums with
associated data, while serde can serialize them

Il dom 14 gen 2024, 12:19 icsmw ***@***.***> ha scritto:

**Comment by icsmw** [comment]
@siefkenj wrote:

> I would like to be able to use automatically generated bindings from wasm modules with typescript and have serialization happen transparently (tsify can mostly do this, but it had some errors with the generated types). wasm-bindgen can automatically convert to JsValue, but the type in the exported .d.ts file is any, rather than the structure of the object created.

Ideally it should not be a some third-party crate... ideally it should be a part of `wasm-bindgen` as soon as it has already at least basic support of TypeScript types... Probably PR to `wasm-bindgen` could be a best solution.
